monero/wallet_client.go

```diff
@@ -221,49 +250,211 @@ func (c *walletClient) WaitForReceipt(req *WaitForReceiptRequest) (*wallet.Trans
+	transfer, err := c.waitForReceipt(&waitForReceiptRequest{
+		Ctx:              ctx,
+		TxID:             reqResp.TxHash,
+		NumConfirmations: numConfirmations,
+		AccountIdx:       accountIdx,
+	})
+	log.Warnf("Transfer TXID=%s failed to get receipt: %s", reqResp.TxHash, err)
+	if err != nil {
+		return nil, fmt.Errorf("monero TXID=%s receipt failure: %w", reqResp.TxHash, err)
+	}
+	log.Infof("Transfer TXID=%s succeeded with %d confirmations and fee %s XMR",
+		transfer.TxID,
+		transfer.Confirmations,
+		coins.FmtPiconeroAmtAsXMR(transfer.Fee),
+	)
+	return transfer, nil
 }
 
-func (c *walletClient) SweepAll(to mcrypto.Address, accountIdx uint64) (*wallet.SweepAllResponse, error) {
-	return c.wRPC.SweepAll(&wallet.SweepAllRequest{
+func (c *walletClient) SweepAll(
+	ctx context.Context,
+	to mcrypto.Address,
+	accountIdx uint64,
+	numConfirmations uint64,
+) ([]*wallet.Transfer, error) {
+	addrResp, err := c.GetAddress(accountIdx)
+	if err != nil {
+		return nil, fmt.Errorf("sweep operation failed to get address: %w", err)
+	}
+	from := addrResp.Address
+	balance, err := c.GetBalance(accountIdx)
+	if err != nil {
+		return nil, fmt.Errorf("sweep operation failed to get balance: %w", err)
+	}
+	log.Infof("Starting sweep of %s XMR from %s to %s", coins.FmtPiconeroAmtAsXMR(balance.Balance), from, to)
+	if balance.BlocksToUnlock > 0 {
+		log.Infof("Sweep operation waiting %d blocks for balance to fully unlock", balance.BlocksToUnlock)
+		if _, err = WaitForBlocks(ctx, c, int(balance.BlocksToUnlock)); err != nil {
+			log.Warnf("Sweep operation from %s failed waiting to unlock balance", from)
+			return nil, fmt.Errorf("sweep operation failed waiting to unlock balance: %w", err)
+		}
+	}
+	if balance.Balance == 0 {
+		log.Warnf("Sweep operation from %s failed, no balance to sweep", from)
+		return nil, errors.New("no balance to sweep")
+	}
+	reqResp, err := c.wRPC.SweepAll(&wallet.SweepAllRequest{
 		AccountIndex: accountIdx,
 		Address:      string(to),
 	})
+	if err != nil {
+		log.Warnf("Sweep transaction(s) from %s failed, %s", from, err)
+		return nil, fmt.Errorf("sweep_all failed: %w", err)
+	}
+	log.Infof("Sweep transaction started, TX IDs: %s", strings.Join(reqResp.TxHashList, ", "))
+	var transfers []*wallet.Transfer
+	for _, txID := range reqResp.TxHashList {
+		receipt, err := c.waitForReceipt(&waitForReceiptRequest{
+			Ctx:              ctx,
+			TxID:             txID,
+			NumConfirmations: numConfirmations,
+			AccountIdx:       accountIdx,
+		})
+		if err != nil {
+			log.Warnf("Sweep TX ID %s failed waiting for receipt: %s", txID, err)
+			return nil, fmt.Errorf("sweep failed waiting for receipt: %w", err)
+		}
+		log.Infof("Sweep transfer ID=%s of %s XMR (%s XMR fees) completed at height %d",
+			txID,
+			coins.FmtPiconeroAmtAsXMR(receipt.Amount),
+			coins.FmtPiconeroAmtAsXMR(receipt.Fee),
+			receipt.Height,
+		)
+		transfers = append(transfers, receipt)
```

Comment:
put this in a go func maybe? might make error handling more annoying, but then it can return on failure immediately if any of the txs fails without having to wait for the earlier ones to complete

Reply:
In all likelihood, there will only be one transaction. Even if there is more than one transaction, it will most likely end up in the same block, so waiting for the 2nd receipt won't block. To prevent `SweepAll` calls from blocking for too long, I made the compromise of setting the number of confirmations to 2 (verses 10 for the transfer that Bob does to the A+B wallet).